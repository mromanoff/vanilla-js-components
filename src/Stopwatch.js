import Component from './Component';

class Stopwatch extends Component {
    constructor(props) {
        super(props);
        this.state = {...props};
        this.formatTime = this.formatTime.bind(this);
        this.start = this.start.bind(this);
        this.stop = this.stop.bind(this);
        this.reset = this.reset.bind(this);
        this.clickHandler = this.clickHandler.bind(this);

        this.timerInterval = null;

        let timer = document.querySelector(this.state.element);
        timer.addEventListener('click', this.clickHandler, false);
    }

    /**
     * Format the time in seconds into hours, minutes, and seconds
     * @param  {Number} time The time in seconds
     * @return {String}      The time in hours, minutes, and seconds
     */
    formatTime(time) {
        let minutes = parseInt(time / 60, 10);
        let hours = parseInt(minutes / 60, 10);
        if (minutes > 59) {
            minutes = minutes % 60;
        }
        return (hours > 0 ? hours + 'h ' : '') + (minutes > 0 || hours > 0 ? minutes + 'm ' : '') + (time % 60) + 's';
    }

    /**
     * Start the stopwatch
     */
    start() {
        let {data: {time}} = this.state;
        // Start the timer
        this.timerInterval = setInterval(() => {
            update_timer()
        }, 1000);
        // Update the timer
        let update_timer = () => this.setState({time: ++time}); // we need access to the current state of time
        this.setState({running: true});
    }

    /**
     * Stop the stopwatch
     */
    stop() {
        this.setState({running: false});
        clearInterval(this.timerInterval);
    }

    /**
     * Reset the stopwatch
     */
    reset() {
        this.setState({time: 0, running: false});
        clearInterval(this.timerInterval);
    }

    clickHandler(event) {
        event.preventDefault();
        // Check if a stopwatch action button was clicked
        let action = event.target.getAttribute('data-stopwatch');
        console.warn(`click action: ${action}`);

        switch (action) {
            case 'start': // If it's the start button, start
                this.start();
                break;
            case 'stop': // If it's the stop button, stop
                this.stop();
                break;
            case 'reset': // If it's the stopwatch button, reset
                this.reset();
                break;
            default:
                return;
        }
    }

    template(props) {
        const {time, running} = props;
        console.warn(props);
        let template = `
        <div id="stopwatch">
          ${this.formatTime(time)}
        </div>
        <div class="stopwatch_content">
          <button 
            data-stopwatch="${ running ? 'stop' : 'start' }"
          > 
            ${ running ? 'Stop' : 'Start' } 
          </button>
          <button data-stopwatch="reset">
            Reset
          </button>
        </div>
      `;
        return template;
    }
};

export default Stopwatch;