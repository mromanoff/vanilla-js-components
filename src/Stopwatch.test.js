/**
 * @jest-environment jsdom
 */

import Stopwatch from './Stopwatch';

xit('Stopwatch renders correctly', () => {

  document.componentRegistry = {};
  document.nextId = 0;

  const element = document.createElement('div');
  element.setAttribute('data-component', 'timer');

  const data = {
    time: 0,
    running: false,
  };

  let component = new Stopwatch({element, data});
  component.render();

  expect(component).toMatchSnapshot();
});