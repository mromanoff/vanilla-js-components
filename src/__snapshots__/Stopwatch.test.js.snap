// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stopwatch renders correctly 1`] = `
Stopwatch {
  "_id": 1,
  "clickHandler": [Function],
  "formatTime": [Function],
  "reset": [Function],
  "start": [Function],
  "state": Object {
    "data": Object {
      "running": false,
      "time": 0,
    },
    "element": <div
      data-component="timer"
    >
      
        
      <div
        class="Stopwatch"
      >
        
          0s
        
      </div>
      
        
      <div
        class="Button-container"
      >
        
          
        <button
          class="Button Button--primary"
          data-stopwatch="start"
        >
           
            Start 
          
        </button>
        
          
        <button
          class="Button"
          data-stopwatch="reset"
        >
          
            Reset
          
        </button>
        
        
      </div>
      
      
    </div>,
  },
  "stop": [Function],
  "timerInterval": null,
}
`;
